unity/EditorPlugin/AfterUnity56/UnitTesting/Initialization.cs

```diff
@@ -4,55 +4,115 @@
+                    throw new InvalidOperationException("Running tests during the Play mode is not possible.");
+                var testLauncher = new UnityEditorTestLauncher(model.UnitTestLaunch.Value, connectionLifetime);
+                return testLauncher.TryLaunchUnitTests();
+            });
+
+            GetUnsavedChangesInScenes(modelAndLifetime);
+        }
+
+        private static void GetUnsavedChangesInScenes(UnityModelAndLifetime modelAndLifetime)
+        {
+            modelAndLifetime.Model.HasUnsavedState.Set(rdVoid =>
+            {
+                var count = SceneManager.sceneCount;
+                for (var i = 0; i < count; i++)
+                {
+                    if (SceneManager.GetSceneAt(i).isDirty)
+                        return true;
+                }
+
+                //Example of ScriptableObject which has its state, independent from the scenes
+                // Add this script to Assets
+                // Create an instance by `Assets > Create > ScriptableObjects > SpawnManagerScriptableObject`
+                // Change SerializableFields in the UnityEditor
+                /* 
+                 [CreateAssetMenu(fileName = "Data", menuName = "ScriptableObjects/SpawnManagerScriptableObject", order = 1)]
+    public class SpawnManagerScriptableObject : ScriptableObject
     {
-        modelAndLifetime.Model.HasUnsavedScenes.Set(rdVoid => 
+        public string prefabName;
+    
+        public int numberOfPrefabsToCreate;
+        public Vector3[] spawnPoints;
+    }
+                 */
+                var hasDirtyUserAssets = Resources.FindObjectsOfTypeAll<Object>()
+                    .Any(a =>
+                    {
+                        var assetPath = AssetDatabase.GetAssetPath(a);
+                        if (string.IsNullOrEmpty(assetPath))
+                            return false;
+                        
+                        if (!IsDirty(a))
+                            return false;
+
+                        return File.Exists(Path.GetFullPath(assetPath));
+                    });
+
+                return hasDirtyUserAssets;
+            });
+        }
+
+        private static bool IsDirty(Object unityObject)
         {
-            var count = SceneManager.sceneCount;
-            for (var i = 0; i < count; i++)
+#if UNITY_2019_2_OR_NEWER
+            return EditorUtility.IsDirty(unityObject);
+#else
+            try
             {
-                if (SceneManager.GetSceneAt(i).isDirty)
-                    return true;
+                var methodInfo = typeof(EditorUtility).GetMethod("IsDirty",
+                    System.Reflection.BindingFlags.Static 
+                    | System.Reflection.BindingFlags.Public 
+                    | System.Reflection.BindingFlags.NonPublic, null, new[] { typeof(int)}, null);
+                if (methodInfo == null)
+                {
+                    ourLogger.Error("IsDirty method not found of type='{0}'", typeof(EditorUtility));
```

Comment:
Will this show in the Unity Console? Do we need it to be an error? We'll be checking this by default, so for (very) old Unity versions, we'd be showing an error that we can't do anything about.

Reply:
It is an error in the log file, not in the Unity Console.